File: app/src/main/java/com/pr0gramm/app/ui/views/viewer/video/TextureViewBackend.java
package com.pr0gramm.app.ui.views.viewer.video;

import android.content.Context;
import android.graphics.SurfaceTexture;
import android.media.MediaPlayer;
import android.view.Surface;
import android.view.TextureView;
import android.view.View;

/**
 * {@link ViewBackend} that uses a {@link TextureView}
 */
class TextureViewBackend implements ViewBackend, TextureView.SurfaceTextureListener {
    private final TextureView textureView;
    private final Callbacks callbacks;

    private Surface surface;

    public TextureViewBackend(Context context, Callbacks callbacks) {
        this.callbacks = callbacks;

        this.textureView = new TextureView(context);
        this.textureView.setSurfaceTextureListener(this);
    }

    @Override
    public void setSize(int width, int height) {
        // do nothing
    }

    @Override
    public void setSurface(MediaPlayer mp) {
        mp.setSurface(surface);
    }

    @Override
    public View getView() {
        return textureView;
    }

    @Override
    public void onSurfaceTextureAvailable(SurfaceTexture surface, int width, int height) {
        this.surface = new Surface(surface);
        callbacks.onAvailable(this);
        callbacks.onSizeChanged(this, width, height);
    }

    @Override
    public void onSurfaceTextureSizeChanged(SurfaceTexture surface, int width, int height) {
        callbacks.onSizeChanged(this, width, height);
    }

    @Override
    public boolean onSurfaceTextureDestroyed(SurfaceTexture surface) {
        callbacks.onDestroy(this);

        if (this.surface != null) {
            this.surface.release();
            this.surface = null;
        }

        return true;
    }

    @Override
    public void onSurfaceTextureUpdated(SurfaceTexture surface) {
    }

    @Override
    public boolean hasSurface() {
        return surface != null;
    }
}

File: app/src/main/java/com/pr0gramm/app/ui/views/viewer/video/TextureViewBackend.kt
package com.pr0gramm.app.ui.views.viewer.video

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.SurfaceTexture
import android.view.Surface
import android.view.TextureView
import android.view.View

/**
 * [ViewBackend] that uses a [TextureView]
 */
internal class TextureViewBackend(
        context: Context, private val callbacks: Callbacks) : TextureView.SurfaceTextureListener {

    private val textureView = TextureView(context).apply {
        surfaceTextureListener = this@TextureViewBackend
    }

    var currentSurface: Surface? = null
        private set

    val view: View = textureView

    @SuppressLint("Recycle")
    override fun onSurfaceTextureAvailable(surface: SurfaceTexture, width: Int, height: Int) {
        currentSurface = Surface(surface)

        callbacks.onAvailable(this)
    }

    override fun onSurfaceTextureSizeChanged(surface: SurfaceTexture, width: Int, height: Int) {
    }

    override fun onSurfaceTextureDestroyed(surface: SurfaceTexture): Boolean {
        callbacks.onDestroy(this)

        this.currentSurface?.release()
        this.currentSurface = null

        return true
    }

    override fun onSurfaceTextureUpdated(surface: SurfaceTexture) {}

    /**
     * Callbacks that are send from a [ViewBackend].
     */
    interface Callbacks {
        fun onAvailable(backend: TextureViewBackend)

        fun onDestroy(backend: TextureViewBackend)
    }
}
